### PR TITLE
[IMP] stock_warehouse_orderpoint_stock_info : fast compute

### PR DESCRIPTION
**Pull request by fmdl**
`_product_available()` methode is compute only once by location.

### CODE REVIEW

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 1]
```diff
@@ -14,14 +14,18 @@ class StockWarehouseOrderpoint(models.Model):
...
-                )._product_available()[rec.product_id.id]
-            rec.product_location_qty = product_available['qty_available']
-            rec.incoming_location_qty = product_available['incoming_qty']
-            rec.outgoing_location_qty = product_available['outgoing_qty']
-            rec.virtual_location_qty = product_available['virtual_available']
+        for loc in self.mapped('location_id'):
+            order_in_loc = self.filtered(lambda x: x.location_id == loc)
+            products = order_in_loc.mapped('product_id').with_context(
+                location=loc.id
+                )._product_available()
```
Comment (lmignon) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: @fmdl You should call `_compute_quantities_dict`in place of `_product_available`. `_product_available`is kept into Odoo for compatibility. https://github.com/odoo/odoo/blob/10.0/addons/stock/models/product.py#L110

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 2]
```diff
@@ -14,14 +14,18 @@ class StockWarehouseOrderpoint(models.Model):
...
     def _compute_product_available_qty(self):
-        for rec in self:
-            product_available = rec.product_id.with_context(
-                location=rec.location_id.id
-                )._product_available()[rec.product_id.id]
-            rec.product_location_qty = product_available['qty_available']
-            rec.incoming_location_qty = product_available['incoming_qty']
-            rec.outgoing_location_qty = product_available['outgoing_qty']
-            rec.virtual_location_qty = product_available['virtual_available']
+        for loc in self.mapped('location_id'):
```
Comment (lmignon) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: @fmdl IMO we should avoid the call to `mapped`followed by a `filtered`. All you need is to process the list of orderpoint by location.
```python
from  collection import defaultdict
op_by_loc = defaultdict(set)
for rec in self:
     op_by_loc(rec.location_id).add(rec)
for location_id, order_in_loc in op_by_loc.items():
    products = order_in_loc.mapped('product_id').with_context(
        location=loc.id)._compute_quantities_dict(lot_id=None, owner_id=None, package_id=None)
    for rec in order_in_loc:
        ...
```

**Review by lmignon — CHANGES_REQUESTED** [review]

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 3]
```diff
@@ -7,21 +7,27 @@
...
     def _compute_product_available_qty(self):
-        for rec in self:
-            product_available = rec.product_id.with_context(
-                location=rec.location_id.id
-                )._product_available()[rec.product_id.id]
-            rec.product_location_qty = product_available['qty_available']
-            rec.incoming_location_qty = product_available['incoming_qty']
-            rec.outgoing_location_qty = product_available['outgoing_qty']
-            rec.virtual_location_qty = product_available['virtual_available']
+        op_by_loc = defaultdict(set)
```
Comment (Garamotte) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: To be able to use `.mapped` later, you have to put data in a recordset, not a set.
`op_by_loc = defaultdict(lambda: self.env['stock.warehouse.orderpoint'])`

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 4]
```diff
@@ -7,21 +7,27 @@
...
+        for order in self:
+            op_by_loc[order.location_id].add(order)
```
Comment (Garamotte) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: In a recordset, you add data by using the `|=` operator.
`op_by_loc[order.location_id] |= order`

Note: The `+=` operator also works, but sometimes adds duplicates in the recordset, that's why I prefer using `|=`.

**Review by Garamotte — CHANGES_REQUESTED** [review]

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py (hunk 4 above)
Comment (lmignon) [reply stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: @sylvain-garancher Good catch. Indeed I hadn't seen the call to `mapped` at line 22

**Review by Garamotte — APPROVED** [review]
Thanks :)

**Review by gurneyalex — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by MiquelRForgeFlow** [comment]
@fmdl, could you squash the commits?

**Comment by fmdl** [comment]
@mreficent what is it squash ?

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 5]
```diff
@@ -7,21 +7,27 @@
...
-            rec.incoming_location_qty = product_available['incoming_qty']
-            rec.outgoing_location_qty = product_available['outgoing_qty']
-            rec.virtual_location_qty = product_available['virtual_available']
+        op_by_loc = defaultdict(lambda: self.env['stock.warehouse.orderpoint'])
+        for order in self:
+            op_by_loc[order.location_id] |= order
+        for location_id, order_in_loc in op_by_loc.items():
+            products = order_in_loc.mapped('product_id').with_context(
+                location=location_id.id)._compute_quantities_dict(
+                lot_id=None, owner_id=None, package_id=None)
```
Comment (rousseldenis) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: @fmdl I read this PR again and I am wondering if these params shouldn't be filled with context values see _compute_quantities() function in stock/models/product.py not to break that behaviour.

**Comment by MiquelRForgeFlow** [comment]
https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

**Comment by fmdl** [comment]
I make the pull request on github website. I'm not an git expert.

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 6]
```diff
@@ -21,7 +21,9 @@ def _compute_product_available_qty(self):
         for location_id, order_in_loc in op_by_loc.items():
             products = order_in_loc.mapped('product_id').with_context(
                 location=location_id.id)._compute_quantities_dict(
-                lot_id=None, owner_id=None, package_id=None)
+                lot_id=self.env.context.get('lot_id'), 
+                owner_id=self.env.context.get('lowner_id'),
```
Comment (rousseldenis) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: Typos in variables but I've deleted my comment. I didn't realized that we are in orderpoints so there is no need of these.

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 7]
```diff
@@ -7,21 +7,29 @@
...
-            rec.virtual_location_qty = product_available['virtual_available']
+        op_by_loc = defaultdict(lambda: self.env['stock.warehouse.orderpoint'])
+        for order in self:
+            op_by_loc[order.location_id] |= order
+        for location_id, order_in_loc in op_by_loc.items():
+            products = order_in_loc.mapped('product_id').with_context(
+                location=location_id.id)._compute_quantities_dict(
+                        lot_id=self.env.context.get('lot_id'), 
+                        owner_id=self.env.context.get('lowner_id'), 
+                        package_id=self.env.context.get('lpackage_id'))
```
Comment (MiquelRForgeFlow) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: 'lowner_id' → 'owner_id'
'lpackage_id' → 'package_id'

But as @rousseldenis said, maybe is not needed these things.

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]

File: stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py [hunk 8]
```diff
@@ -7,21 +7,29 @@
...
-            rec.virtual_location_qty = product_available['virtual_available']
+        op_by_loc = defaultdict(lambda: self.env['stock.warehouse.orderpoint'])
+        for order in self:
+            op_by_loc[order.location_id] |= order
+        for location_id, order_in_loc in op_by_loc.items():
+            products = order_in_loc.mapped('product_id').with_context(
+                location=location_id.id)._compute_quantities_dict(
+                        lot_id=self.env.context.get('lot_id'), 
+                        owner_id=self.env.context.get('owner_id'), 
+                        package_id=self.env.context.get('package_id'))
```
Comment (MiquelRForgeFlow) [inline stock_warehouse_orderpoint_stock_info/models/stock_warehouse_orderpoint.py]: selection_024

Please, delete this white spaces

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by LoisRForgeFlow** [comment]
@lmignon Could you update your review?

**Comment by LoisRForgeFlow** [comment]
@lmignon you review is blocking, could you please update it?

**Review by lmignon — APPROVED** [review]

**Comment by lmignon** [comment]
@lreficent Done! Sorry for the delay

**Pull request merged**